joey/layers_v2.py

```diff
@@ -391,3 +393,357 @@ def __init__(self, kernel_size, input_size,
+                         alloc, dim_alloc,
+                         generate_code)
+
+    def _set_padding_result_values(self, input_func, result_func, value=0):
+        devito_func_dims = input_func.dimensions[2:]
+        eqs = []
+        for i, dim in enumerate(devito_func_dims):
+            dim_l = SubDimension.left(name='sub%s_l' % dim.name, parent=dim,
+                                      thickness=self._padding[i])
+            eqs.append(Inc(input_func.subs({dim: dim_l}), value))
+            dim_r = SubDimension.right(name='sub%s_r' % dim.name, parent=dim,
+                                       thickness=self._padding[i])
+            eqs.append(Inc(input_func.subs({dim: dim_r}), value))
+        eqs.append(Eq(result_func, value))
+        op = Operator(eqs)
+        op.apply()
+
+    def _error_check(self, kernel_size, input_size, stride, padding,
+                     strict_stride_check):
+        if input_size is None or (len(input_size) != self._dims+2):
+            raise Exception("Input size is incorrect")
+
+        if kernel_size is None or (len(kernel_size) != self._dims):
+            raise Exception("Kernel size is incorrect")
+
+        if stride is None or (len(stride) != self._dims):
+            raise Exception("Stride is incorrect")
+
+        if padding is None or (len(padding) != self._dims):
+            raise Exception("Padding is incorrect")
+
+        for i in range(0, self._dims):
+
+            if stride[i] < 1:
+                raise Exception("Stride cannot be less than 1")
+
+            if padding[i] < 0:
+                raise Exception("Padding cannot be negative")
+
+        if strict_stride_check:
+            input_d = input_size[-self._dims+i] + 2 * padding[i]
+            if (input_d - kernel_size[-self._dims+i]) % stride[i] != 0:
+                raise Exception("Stride " + str(stride) + " is not "
+                                "compatible with feature map, kernel and "
+                                "padding sizes. If you want to proceed "
+                                "anyway, set strict_stride_check=False "
+                                "when instantiating this object")
+
+    def _allocate(self, kernel_size, input_size, name_allocator_func,
+                  dim_allocator_func):
+
+        dimensions = ['dbatch', 'dchannel']
+        result_shape = []
+        input_size = list(input_size)
+        # generating  in the order depth, height, width ,
+        # hence arr[-3], arr[-2] and so on
+        for i in range(0, self._dims):
+            result_d = (input_size[(-self._dims+i)] -
+                        kernel_size[(-self._dims+i)] +
+                        2 * self._padding[i])//self._stride[i] + 1
+            result_shape.append(result_d)
+            input_size[(-self._dims+i)] += 2 * self._padding[i]
+            dimensions.append(
+                'd_'+self._dim_dict.get(self._dims-i, self._dims-i))
```

Comment:
This can be rewritten in a way that is much easier to read, if the names depth, height, width were used.

Reply:
So if we were to extend this to Nd pooling I wasn't sure of the naming convention. I will rethink and simplify this.

Reply:
Agreed that this change is not necessary. However, `self._dims` should be renamed `self._ndims` to make it clear that it is an integer.